lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp

```diff
@@ -184,18 +112,21 @@ bool ExistentialSpecializer::canSpecializeExistentialArgsInFunction(
       continue;
 
     auto ExistentialRepr =
-        Arg->getType().getPreferredExistentialRepresentation(F->getModule());
+        CalleeArg->getType().getPreferredExistentialRepresentation(
+            F->getModule());
     if (ExistentialRepr != ExistentialRepresentation::Opaque &&
           ExistentialRepr != ExistentialRepresentation::Class)
       continue;
 
     /// Find the concrete type.
-    CanType ConcreteType;
-    if (!findConcreteType(Apply, Idx, ConcreteType)) {
+    Operand &ArgOper = Apply.getArgumentRef(Idx);
+    CanType ConcreteType =
+        ConcreteExistentialInfo(ArgOper.get(), ArgOper.getUser()).ConcreteType;
```

Comment:
Thanks a lot for doing this. This was due from my side as the next PR cleanup :) You should also look at this other PR that also tries to create a concrete existential using SoleConformingType: https://github.com/apple/swift/pull/20977

Ideally, we should do all these in Existential.cpp with both ExistentialSpecializer and SILCombiner as consumers. Then we would have to pass PCA and CHA to this function.

Reply:
The PRs look mostly complementary. In this PR, the ConcreteExistentialInfo constructor always first attempts to find an init_existential first before using the provided ConreteType--it basically avoids doing an unnecessary cast. If you end up merging first, I can change this PR so that it's the caller's job to first attempt to create CEI from an init_existential, then create a different CEI using the sole conforming type.